Add Snakemake as a dependency for the Nextstrain environment

huddlej opened the pull request:
Maintains presence of snakemake executable in the Nextstrain environment in
preparation for removing Snakemake as an augur dependency [1]. Two important
points to note about this change are:

  1. We install Snakemake from PyPI instead of conda to ensure we have the
  latest version and to minimize conda environment conflicts that have been
  issues for many users in the past.

  2. We do not pin Snakemake to a specific earlier version here as we did in
  augur (v5.10.0). This means the latest version of Snakemake will be installed
  by default with whatever breaking changes might occur in Snakemake's own
  API. This allows us to take advantage of the the latest features and bug fixes
  for Snakemake, but it also creates a breaking change for augur that will be
  reflected in a major version release of augur.

[1] https://github.com/nextstrain/augur/pull/557

tsibley commented:
> We install Snakemake from PyPI instead of conda to ensure we have the
latest version and to minimize conda environment conflicts that have been
issues for many users in the past.

Can you say more about the circumstances around Conda environment conflicts re: Snakemake? Not familiar with these, and I'm curious to understand more about the issue.

huddlej commented:
This took me a while to dig into because I never kept notes on these issues despite running into them multiple times. There are actually a couple of issues that have made conda installation of Snakemake painful. 

1. Snakemake's dependencies are sourced from the conda-forge channel. This channel is notoriously slow to load and parse and we've tried to avoid it as much as possible for this reason. In writing this up, I added conda-forge to the list of channels in the YAML and moved `snakemake` from pip to the top-level conda dependencies and tried to create the Nextstrain environment: 10 minutes later it is still "Solving environment". Without conda-forge, the installation takes about a minute. (30 minutes later, still solving the environment. I'm cancelling it, since I can't bother to wait.)

2. Installing Snakemake with conda as recommended in the docs can result in an older version being installed, as described in [an issue in the conda repo](https://github.com/conda/conda/issues/9905). When you would try to force the installation of snakemake with the later version, conda would complain that it couldn't solve the environment for that version.

[Snakemake docs now recommend using mamba for installation to avoid these issues](https://snakemake.readthedocs.io/en/stable/getting_started/installation.html#installation-via-conda). Mamba is **way** faster, resolving the environment in less than a minute. It still installed snakemake 5.14.0 even though the latest version in PyPI is 5.19.3. When I requested that mamba install snakemake=5.19.3, this worked fine.

I’m probably going to switch to using mamba for my personal use after this experience, but I don’t think we should push Nextstrain users in that direction just for snakemake. Installing from pip is easy enough, it allows us to install snakemake quickly, and it always grabs the latest version like we’d expect.

tsibley commented:
That is super helpful and informative, John. Thank you so much!

emmahodcroft commented:
Yes, this is super thorough John, thank you! Just so I know I'm understanding clearly, so the proposal would be to remove it from the `augur` conda environment/install, and instead guide users to just install it via pip (in that environment)? That seems like it should work fine :)

tsibley commented:
> Just so I know I'm understanding clearly, so the proposal would be to remove it from the augur conda environment/install, and instead guide users to just install it via pip (in that environment)? That seems like it should work fine :)

Not quite. The change is removing Snakemake from augur's `setup.py` dependencies and adding to the Nextstrain Conda environment's dependencies here (which is used to publish http://data.nextstrain.org/nextstrain.yml). Snakemake is still installed by Conda automatically when creating the environment, but it's installed by Conda via pip packages instead of from Conda's own packages.

huddlej commented:
@tsibley I tested this out a bit more locally and confirmed that the snakemake installed by this conda environment will override the pinned earlier version required by augur.

One potential issue is that this PR will change the behavior of snakemake (mostly making `--cores` required) and that we don't document that change here. We could wait until docs are updated in augur/nextstrain.org, but I'm leaning toward merging this now so we can move ahead with removing the snakemake dependency in augur.

emmahodcroft commented:
@tsibley Thank you! That sounds even better :)

@huddlej It might be good to add this even in just a couple places (and maybe to our `ncov` instructions?) just so people don't get too confused. @eharkins I'm a bit behind in the moving of the docs... where would I go to edit this quickly for install/instructions - is it still all nextstrain.org? John, maybe you could add a tiny bit somewhere at the start of `ncov` tutorial? (I'm hoping saying once is enough.) Then we can go ahead with a clear-conscience! 😁

huddlej commented:
I added `--cores` to the only Snakemake command in our docs that didn't already specify cores through a profile and also added an example to [the running guide](https://nextstrain.github.io/ncov/running) of how to specify `--cores` from the command line to override the profile's value.

After making these edits, I can see it makes more sense to 1) update all docs to use `snakemake --cores`, 2) merge this PR, and then 3) merge the augur PR removing the Snakemake dependency. This should break the fewest things for people.

emmahodcroft commented:
Thanks John! This sounds like a good plan :D

eharkins commented:
@emmahodcroft Sorry I'm not sure I understand the docs question here:

>  huddlej It might be good to add this even in just a couple places (and maybe to our ncov instructions?) just so people don't get too confused. @eharkins I'm a bit behind in the moving of the docs... where would I go to edit this quickly for install/instructions - is it still all nextstrain.org? 

Was it about where we would need to add in `--cores` to snakemake commands in docs?

I'm happy to try and help with the docs side of @huddlej's [proposed plan](https://github.com/nextstrain/conda/pull/8#issuecomment-653668986) if anything still needs doing!

emmahodcroft commented:
Sorry for not being more clear @eharkins ! Yes - I just meant, if we want to add this to the tutorials (TB & Zika) + other associated docs (maybe there's a line in install, I can't remember) is this still all living in `nextstrain.org`, or is it somewhere else now? I've seen some stuff about docs moving, but can't quite figure out what's pulled from where...

eharkins commented:
@emmahodcroft Ah sorry yeah no big shifts in terms of docs hosting as of yet. Zika and tb tutorials are still on nextstrain.org as well as augur docs. So I would imagine now that John updated augur docs to there may still be related changes to make in nextstrain.org tutorials/ how-tos etc and perhaps other places (see [this thread on slack](https://bedfordlab.slack.com/archives/C7SDVPBLZ/p1590556700158900) for a link to the google doc attempting to track where docs currently live).

huddlej commented:
@eharkins @emmahodcroft @tsibley We should be ready to merge this PR now. I've updated the augur and nextstrain.org docs, as discussed above, to reflect the requirement of `--cores` and also mention when Snakemake needs to be installed separately.

If you're interested, I've made [a checklist for the final removal of the Snakemake dependency from augur](https://github.com/nextstrain/augur/pull/557#issuecomment-672293573) that shows where we are in the unexpectedly complicated process.

huddlej commented:
I've tested this branch on the zika-tutorial and ncov repos without any issues. Merging now.
